packages/scm/src/browser/dirty-diff/diff-computer.ts

```diff
@@ -24,21 +25,25 @@ export class DiffComputer {
         return diffResult;
     }
 
-    computeDirtyDiff(previous: ContentLinesArrayLike, current: ContentLinesArrayLike): DirtyDiff {
+    computeDirtyDiff(previous: ContentLinesArrayLike, current: ContentLinesArrayLike, options?: DirtyDiffOptions): DirtyDiff {
         const added: LineRange[] = [];
         const removed: number[] = [];
         const modified: LineRange[] = [];
+        const rangeMappings: ChangeRangeMapping[] | undefined = options?.rangeMappings ? [] : undefined;
         const changes = this.computeDiff(previous, current);
-        let lastLine = -1;
+        let currentRevisionLine = -1;
+        let previousRevisionLine = -1;
```

Comment:
I'm sorry, but I don't see how the suggested algorithm is simpler or more efficient or even much different than what is currently implemented.

For example,

```
	if (change.type === unchanged) {
	  readIndex+= change.count; 
	  writeIndex+= change.count;
	}
```

is the proposal precisely matches

```
	} else {
		// case: unchanged region
		currentRevisionLine += change.count!;
		previousRevisionLine += change.count!;
	}

```

in the current implementation, i.e. `readIndex` exactly corresponds to `previousRevisionLine`, and `writeIndex` exactly corresponds to `currentRevisionLine` in this and all the other cases.

Also, I don't think that ignoring `values` would make much if any difference. Essentially, it would be nothing more than recomputing what has already been computed.

The only significant difference that I can see is that the current implementation uses "look ahead" for coalescing, while the suggested algorithm uses "mutation" for that. That is, instead of

```
	const change = diffResult[i];
	const next = diffResult[i + 1];
	...
	} else if (change.removed && next && next.added) {
		...
		changes.push({ previousRange: toLineRange(change), currentRange: toLineRange(next) });
		currentRevisionLine += next.count!;
		previousRevisionLine += change.count!;
		i++; // consume next eagerly
```

in the current implementation, the suggested algorithm uses something like

```
	} else if (change.type === added) {
	  if (lastRangeMapping) {
		lastRangeMapping.newRange.end += change.count;
		if (lastRangeMapping?.type === removed) {
		  lastRangeMapping.type = modified;
		}
	  }
	  ...
	  writeIndex += change.count;
	}
```

I don't think that there is much difference in simplicity or efficiency between these two approaches essentially.

But I do think that the "look ahead" approach is a better fit in this particular case:

- It aligns well with the immutability of the dirty diff structures, including `LineRange`, which I believe is a good design.

- It is used not only for coalescing, but also for handling special cases such as removing at the beginning or end of the file (see references to `isNextEmptyLine`).

By the way, it is these special cases that probably make the current implementation look more complex that it actually is. And I don't think that the suggested approach would make handling of these special cases any simpler.

Finally, *if it ain't broke, don't fix it*. And I believe it ain't broke.

Reply:
Well the point is that the special cases would go away, but I won't belabour said point.